### PR TITLE
Query Execution: Better exception handling on unawaited async tasks

### PR DESCRIPTION
**Pull request by benrr101**
This change adds:
* A new extension method to Task that adds a ContinueWith to the task that will log any and all exceptions. It will optionally call a provided action, as well.
* Utilization of the new ContinueWithOnFaulted in query execution (not simple)
* Utilization of the new ContinueWithOnFaulted in query save as
* Unit tests for the ContinueWithOnFaulted code
* Some light code cleanup in Query class and UnitTest.csproj

### CODE REVIEW

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/Query.cs [hunk 1]
```diff
@@ -269,7 +276,11 @@ public void Cancel()
         /// </summary>
         public void Execute()
         {
-            ExecutionTask = Task.Run(ExecuteInternal);
+            ExecutionTask = Task.Run(ExecuteInternal)
+                .ContinueWithOnFaulted(t =>
+                {
+                    QueryFailed?.Invoke(this, t.Exception).Wait();
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/QueryExecution/Query.cs]: Minor: should we trace too?

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/Query.cs [hunk 2]
```diff
@@ -405,16 +418,18 @@ public Task<ExecutionPlan> GetExecutionPlan(int batchIndex, int resultSetIndex)
             }
             finally
             {
+                // Remove the message hadler from the connection
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/QueryExecution/Query.cs]: typo: `hadler` -> `handler`

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs [hunk 3]
```diff
@@ -504,9 +504,15 @@ public void RemoveRow(long internalId)
                     }
                 }
             });
+            
+            // Add exception handling to the save task
+            Task taskWithHandling = saveAsTask.ContinueWithOnFaulted(t =>
+            {
+                failureHandler?.Invoke(saveParams, t.Exception.Message);
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs]: There's no Wait() needed on this one?

Comment (kevcunnane) [reply src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs]: You're awaiting it elsewhere.

File: src/Microsoft.SqlTools.ServiceLayer/Utility/TaskExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        /// </param>
+        /// <returns>Task with exception handling on continuation</returns>
+        public static Task ContinueWithOnFaulted(this Task task, Action<Task> continuationAction)
+        {
+            return task.ContinueWith(t =>
+            {
+                // If the task hasn't faulted or has an exception, skip processing
+                if (!t.IsFaulted || t.Exception == null)
+                {
+                    return;
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/Utility/TaskExtensions.cs]: Minor: we don't currently have a case where we need to support this on a task with a return type?

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/Query.cs (hunk 1 above)
Comment (kevcunnane) [reply src/Microsoft.SqlTools.ServiceLayer/QueryExecution/Query.cs]: Ah - I see you're logging in the extension method. Never mind.

File: test/Microsoft.SqlTools.ServiceLayer.UnitTests/Utility/TaskExtensionTests.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace Microsoft.SqlTools.ServiceLayer.UnitTests.Utility
+{
+    public class TaskExtensionTests
+    {
+        [Fact]
+        public async Task ContinueWithOnFaultedNullContinuation()
+        {
+            // Setup: Create a task that will definitely fault
+            Task failuretask = new Task(() => throw new Exception("It fail!"));
```
Comment (kevcunnane) [inline test/Microsoft.SqlTools.ServiceLayer.UnitTests/Utility/TaskExtensionTests.cs]: Minor: `failureTask` (camelcase)

File: src/Microsoft.SqlTools.ServiceLayer/Utility/TaskExtensions.cs (hunk 4 above)
Comment (benrr101) [reply src/Microsoft.SqlTools.ServiceLayer/Utility/TaskExtensions.cs]: I don't think so. If the point of launching a task asynchronously we probably wouldn't ever need it to return something, so just adding onto Task should be fine.

File: src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs (hunk 3 above)
Comment (benrr101) [reply src/Microsoft.SqlTools.ServiceLayer/QueryExecution/ResultSet.cs]: I think so...?

**Pull request merged**